x-pack/legacy/plugins/lens/public/lens_ui_telemetry/factory.ts

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import moment from 'moment';
+import { HttpServiceBase } from 'src/core/public';
+
+import { Storage } from 'src/legacy/core_plugins/data/public/types';
+import { BASE_API_URL } from '../../common';
+
+const STORAGE_KEY = 'lens-ui-telemetry';
+
+let reportManager: LensReportManager;
+
+export function setReportManager(newManager: LensReportManager) {
+  if (reportManager) {
+    reportManager.stop();
+  }
+  reportManager = newManager;
+}
+
+export function stopReportManager() {
+  if (reportManager) {
+    reportManager.stop();
+  }
+}
+
+export function trackUiEvent(name: string) {
+  if (reportManager) {
+    reportManager.trackEvent(name);
+  }
+}
+
+export function trackSuggestionEvent(name: string) {
+  if (reportManager) {
+    reportManager.trackSuggestionEvent(name);
+  }
+}
+
+export class LensReportManager {
+  private events: Record<string, Record<string, number>> = {};
+  private suggestionEvents: Record<string, Record<string, number>> = {};
+
+  private storage: Storage;
+  private http: HttpServiceBase;
+  private basePath: string;
+  private timer: ReturnType<typeof setInterval>;
+
+  constructor({
+    storage,
+    http,
+    basePath,
+  }: {
+    storage: Storage;
+    http: HttpServiceBase;
+    basePath: string;
+  }) {
+    this.storage = storage;
+    this.http = http;
+    this.basePath = basePath;
+
+    this.readFromStorage();
+
+    this.timer = setInterval(() => {
+      this.postToServer();
+    }, 10000);
+  }
+
+  public trackEvent(name: string) {
+    this.readFromStorage();
+    this.trackTo(this.events, name);
+  }
+
+  public trackSuggestionEvent(name: string) {
+    this.readFromStorage();
+    this.trackTo(this.suggestionEvents, name);
+  }
+
+  public stop() {
+    if (this.timer) {
+      clearInterval(this.timer);
+    }
+  }
+
+  private readFromStorage() {
+    const data = this.storage.get(STORAGE_KEY);
+    if (data && typeof data.events === 'object' && typeof data.suggestionEvents === 'object') {
+      this.events = data.events;
+      this.suggestionEvents = data.suggestionEvents;
+    }
+  }
+
+  private async postToServer() {
+    this.readFromStorage();
```

Comment:
This is probably fine, but it does introduce a disk read for a whole bunch of user-driven events. I wonder if this'll have any impact on performance on low-end devices... Probably not.

Anyway, an alternative would be to track all the stuff in memory, but when you flush it to storage, read and merge then, and reset your in-memory state.

Reply:
There are probably some alternatives here, but this is the simplest implementation of it